tests/test_postfix.py

```diff
@@ -0,0 +1,366 @@
+    container_per_test: ContainerData,
+) -> None:
+    # test sending an email successfully to virtual mailbox, using Postfix container
+
+    assert container_per_test.connection.run_expect(
+        [0], "chown -R vmail:vmail /var/spool/vmail/"
+    )
+    assert container_per_test.connection.run_expect(
+        [0], "nohup /usr/local/bin/mailhog > /dev/null 2>&1 &"
+    )
+
+    sendmail_cmd = 'echo "Subject: Test Email\n\nThis is a test email body." | sendmail -f user1@example.com user2@example.com'
+    assert container_per_test.connection.run_expect([0], sendmail_cmd)
+
+    assert "Mail queue is empty" in container_per_test.connection.check_output(
+        "mailq"
+    )
+
+    for output in [
+        "from=<user1@example.com>",
+        "to=<user2@example.com>",
+        "relay=virtual",
+        "dsn=2.0.0",
+        "status=sent (delivered to maildir)",
+    ]:
+        assert output in container_per_test.connection.check_output(
+            "cat /var/log/postfix.log"
+        )
+
+    assert container_per_test.connection.file(
+        "/var/spool/vmail/example.com/user2"
+    ).exists
+
+
+if OS_VERSION == "tumbleweed":
+    CONTAINERFILE_POSTFIX_WITH_LDAP_ENABLED = """
+RUN set -euo pipefail; \
+zypper -n in --no-recommends git find openldap2 openldap2-client; \
+zypper -n clean; \
+rm -rf /var/log/{lastlog,tallylog,zypper.log,zypp/history,YaST2}
+
+# TODO: move postfix & openldap container files to bci-dockerfile-generator
+RUN git clone https://github.com/thkukuk/containers-mailserver.git && \
+        cd containers-mailserver/openldap && \
+        cp -r ldif /entrypoint/ && \
+        cp slapd.init.ldif /entrypoint/ && \
+        cp entrypoint.sh /entrypoint/openldap-entrypoint.sh
+
+RUN FILE="/entrypoint/openldap-entrypoint.sh" && \
+        for cmd in \
+            's|-h "$LDAP_URL $LDAPS_URL $LDAPI_URL" ${SLAPD_SLP_REG}|-h "$LDAP_URL $LDAPS_URL $LDAPI_URL" ${SLAPD_SLP_REG} > /var/log/slapd.log 2>\\&1 \\&|' \
+            's|exec /usr/sbin/slapd -d "${SLAPD_LOG_LEVEL}" -u ldap -g ldap|exec nohup /usr/sbin/slapd -d "${SLAPD_LOG_LEVEL}" -u ldap -g ldap|' \
+            's|ldapadd -c -Y EXTERNAL -Q -H ldapi:/// -f /etc/openldap/schema/ppolicy.ldif||' \
+        ; do \
+            sed -i "$cmd" "$FILE"; \
+        done
+
+RUN cat <<EOF > /entrypoint/ldif/examples/example-user.ldif
+dn: uid=user1,ou=mail,dc=example,dc=com
+cn: user1
+gidnumber: 20001
+homedirectory: /home/mail/user1
+mailacceptinggeneralid: user1@example.com
+maildrop: user1@mail.example.com
+objectclass: account
+objectclass: posixAccount
+objectclass: postfixUser
+objectclass: top
+uid: user1
+uidnumber: 20001
+userpassword: user1
+
+dn: uid=user2,ou=mail,dc=example,dc=com
+cn: user2
+gidnumber: 20002
+homedirectory: /home/mail/user2
+mailacceptinggeneralid: user2@example.com
+maildrop: user2@mail.example.com
+objectclass: account
+objectclass: posixAccount
+objectclass: postfixUser
+objectclass: top
+uid: user2
+uidnumber: 20002
+userpassword: user2
+EOF
```

Comment:
I'm afraid buildah & docker will not accept this, you'll have to resort to the much uglier ANSI-C quoting: https://stackoverflow.com/a/33439625
```suggestion
RUN echo $'dn: uid=user1,ou=mail,dc=example,dc=com \n\
cn: user1 \n\
gidnumber: 20001 \n\
homedirectory: /home/mail/user1 \n\
mailacceptinggeneralid: user1@example.com \n\
maildrop: user1@mail.example.com \n\
objectclass: account \n\
objectclass: posixAccount \n\
objectclass: postfixUser \n\
objectclass: top \n\
uid: user1 \n\
uidnumber: 20001 \n\
userpassword: user1 \n\
\n\
dn: uid=user2,ou=mail,dc=example,dc=com \n\
cn: user2 \n\
gidnumber: 20002 \n\
homedirectory: /home/mail/user2 \n\
mailacceptinggeneralid: user2@example.com \n\
maildrop: user2@mail.example.com \n\
objectclass: account \n\
objectclass: posixAccount \n\
objectclass: postfixUser \n\
objectclass: top \n\
uid: user2 \n\
uidnumber: 20002 \n\
userpassword: user2 '  > /entrypoint/ldif/examples/example-user.ldif
```

Reply:
@dcermak, the heredocs format is working for me in my local testing with podman/buildah.
Just verified also for official support in buildah & docker.

The [heredocs support in buildah](https://github.com/containers/buildah/pull/5092) is available as of 1.33.0 - https://buildah.io/releases/2023/11/17/Buildah-version-v1.33.0.html#changes-for-v1330

Docker also seems to support it - https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/

---

With the ANSI-C quoting format, while testing locally, I'm getting error for `opening quote not matching the closing quote` (because of multi-line content) in Containerfile. Tried with changing shells as well for RUN (sh to bash), but same.

```
/bin/sh: -c: line 1: unexpected EOF while looking for matching ''
Error: building at STEP "RUN echo $'dn: uid=user1,ou=mail,dc=example,dc=com": while running runtime: exit status 2
```

And when I test the ANSI-C format locally, another problem is the multi-line echo actually keeps the content on multiline, and then the new line `\n` creates additional new line, and the backslash gets printed too.

```
dn: uid=user1,ou=mail,dc=example,dc=com 
\
cn: user1 
\
gidnumber: 20001 
\
homedirectory: /home/mail/user1 
\
mailacceptinggeneralid: user1@example.com 
\
maildrop: user1@mail.example.com 
\
objectclass: account 
\
objectclass: posixAccount 
\
objectclass: postfixUser 
\
objectclass: top 
\
uid: user1 
\
uidnumber: 20001 
\
userpassword: user1 
\

\
dn: uid=user2,ou=mail,dc=example,dc=com 
\
cn: user2 
\
gidnumber: 20002 
\
homedirectory: /home/mail/user2 
\
mailacceptinggeneralid: user2@example.com 
\
maildrop: user2@mail.example.com 
\
objectclass: account 
\
objectclass: posixAccount 
\
objectclass: postfixUser 
\
objectclass: top 
\
uid: user2 
\
uidnumber: 20002 
\
userpassword: user2 
```